File: technical/structures.py
'''
Arrays

Create the following arrays

unsorted
5, 8, 4, 6, 2, 9, 1, 3, 7

sorted
1, 2, 3, 4, 5, 6, 7, 8, 9

'''

arrayOne = [5, 8, 4, 6, 2, 9, 1, 3, 7]
sortedArrayOne = [1, 2, 3, 4, 5, 6, 7, 8, 9]


'''
Trees
'''
class TreeNode:
    def __init__(self, n):
        self.val = n
        self.left = None
        self.right = None


'''

Create the following BST

        7
    3       11
1     4  9      13
 2

'''

bstOneRoot = TreeNode(7)
bstOneRoot.left = TreeNode(3)
bstOneRoot.right = TreeNode(11)
bstOneRoot.left.left = TreeNode(1)
bstOneRoot.left.left.right = TreeNode(2)
bstOneRoot.left.right = TreeNode(4)
bstOneRoot.right.left = TreeNode(9)
bstOneRoot.right.right = TreeNode(13)


'''
Graphs
'''
class GraphNode:
    def __init__(self, n):
        self.val = n
        self.neighbors = []

'''

Create the following Directed Graphs

dagOne
0 -> 4, 5
1 -> 2, 5
2 -> 3
3 -> 4
4 -> 
5 ->

dagTwo
0 -> 1
1 -> 2
2 -> 3
3 -> 4
4 -> 5
5 ->

dgWCycle
0 -> 1
1 -> 2
2 -> 3, 4
3 ->
4 -> 5
5 -> 2

'''

dagOne = {}
dagOne[0] = [4, 5]
dagOne[1] = [2, 5]
dagOne[2] = [3]
dagOne[3] = [4]
dagOne[4] = []
dagOne[5] = []

dagTwo = {}
dagTwo[0] = [1]
dagTwo[1] = [2]
dagTwo[2] = [3]
dagTwo[3] = [4]
dagTwo[4] = [5]
dagTwo[5] = []

dgWCycle = {}
dgWCycle[0] = [1]
dgWCycle[1] = [2]
dgWCycle[2] = [3, 4]
dgWCycle[3] = []
dgWCycle[4] = [5]
dgWCycle[5] = [2]


'''

Create the following Undirected Graphs

ugOne
0 -> 1
1 -> 0, 2
2 -> 1, 3
3 -> 2, 4
4 -> 3

ugWCycle
0 -> 1
1 -> 0, 2
2 -> 1, 3, 4, 5
3 -> 2
4 -> 2, 5
5 -> 2, 4

'''

ugOne = {}
ugOne[0] = [1]
ugOne[1] = [0, 2]
ugOne[2] = [1, 3]
ugOne[3] = [2, 4]
ugOne[4] = [3]

ugWCycle = {}
ugWCycle[0] = [1]
ugWCycle[1] = [0, 2]
ugWCycle[2] = [1, 3, 4, 5]
ugWCycle[3] = [2]
ugWCycle[4] = [2, 5]
ugWCycle[5] = [2, 4]


'''
Linked Lists
'''
class SingleLLNode:
    def __init__(self, n):
        self.val = n
        self.next = None

'''

Create the following singly linked list

4 -> 7 -> 3 -> 9 -> 1 -> 8

'''

sllOneHead = SingleLLNode(4)
sllOneHead.next = SingleLLNode(7)
sllOneHead.next.next = SingleLLNode(3)
sllOneHead.next.next.next = SingleLLNode(9)
sllOneHead.next.next.next.next = SingleLLNode(1)
sllOneHead.next.next.next.next.next = SingleLLNode(8)

class DoubleLLNode:
    def __init__(self, n):
        self.val = n
        self.next = None
        self.prev = None

File: technical-interview/structures.rs
use std::collections::HashMap;

/*
Arrays

Create the following arrays

unsorted
5, 8, 4, 6, 2, 9, 1, 3, 7

sorted
1, 2, 3, 4, 5, 6, 7, 8, 9
*/

pub fn array_one() -> Vec<i32> {
    vec![5, 8, 4, 6, 2, 9, 1, 3, 7]
}

pub fn sorted_array_one() -> Vec<i32> {
    vec![1, 2, 3, 4, 5, 6, 7, 8, 9]
}

/*
Trees
*/

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Box<TreeNode>>,
    pub right: Option<Box<TreeNode>>,
}

impl TreeNode {
    pub fn new(val: i32) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }
}

/*
Create the following BST

        7
    3       11
1     4  9      13
 2
*/

pub fn bst_one_root() -> Option<Box<TreeNode>> {
    let mut root = Box::new(TreeNode::new(7));

    let mut left = Box::new(TreeNode::new(3));
    let mut left_left = Box::new(TreeNode::new(1));
    left_left.right = Some(Box::new(TreeNode::new(2)));
    left.left = Some(left_left);
    left.right = Some(Box::new(TreeNode::new(4)));

    let mut right = Box::new(TreeNode::new(11));
    right.left = Some(Box::new(TreeNode::new(9)));
    right.right = Some(Box::new(TreeNode::new(13)));

    root.left = Some(left);
    root.right = Some(right);

    Some(root)
}

/*
Graphs

Create the following Directed Graphs

dagOne
0 -> 4, 5
1 -> 2, 5
2 -> 3
3 -> 4
4 ->
5 ->

dagTwo
0 -> 1
1 -> 2
2 -> 3
3 -> 4
4 -> 5
5 ->

dgWCycle
0 -> 1
1 -> 2
2 -> 3, 4
3 ->
4 -> 5
5 -> 2
*/

pub type Graph = HashMap<i32, Vec<i32>>;

pub fn dag_one() -> Graph {
    HashMap::from([
        (0, vec![4, 5]),
        (1, vec![2, 5]),
        (2, vec![3]),
        (3, vec![4]),
        (4, vec![]),
        (5, vec![]),
    ])
}

pub fn dag_two() -> Graph {
    HashMap::from([
        (0, vec![1]),
        (1, vec![2]),
        (2, vec![3]),
        (3, vec![4]),
        (4, vec![5]),
        (5, vec![]),
    ])
}

pub fn directed_graph_with_cycle() -> Graph {
    HashMap::from([
        (0, vec![1]),
        (1, vec![2]),
        (2, vec![3, 4]),
        (3, vec![]),
        (4, vec![5]),
        (5, vec![2]),
    ])
}

/*
Create the following Undirected Graphs

ugOne
0 -> 1
1 -> 0, 2
2 -> 1, 3
3 -> 2, 4
4 -> 3

ugWCycle
0 -> 1
1 -> 0, 2
2 -> 1, 3, 4, 5
3 -> 2
4 -> 2, 5
5 -> 2, 4
*/

pub fn undirected_graph_one() -> Graph {
    HashMap::from([
        (0, vec![1]),
        (1, vec![0, 2]),
        (2, vec![1, 3]),
        (3, vec![2, 4]),
        (4, vec![3]),
    ])
}

pub fn undirected_graph_with_cycle() -> Graph {
    HashMap::from([
        (0, vec![1]),
        (1, vec![0, 2]),
        (2, vec![1, 3, 4, 5]),
        (3, vec![2]),
        (4, vec![2, 5]),
        (5, vec![2, 4]),
    ])
}

/*
Linked Lists
*/

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct SingleLLNode {
    pub val: i32,
    pub next: Option<Box<SingleLLNode>>,
}

impl SingleLLNode {
    pub fn new(val: i32) -> Self {
        Self { val, next: None }
    }
}

#[allow(dead_code)]
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DoubleLLNode {
    pub val: i32,
    pub next: Option<Box<DoubleLLNode>>,
    // A fully-owned doubly linked list in Rust is usually built with Rc<RefCell<T>>
    // or raw pointers. Keep this placeholder simple for interview practice.
}

impl DoubleLLNode {
    #[allow(dead_code)]
    pub fn new(val: i32) -> Self {
        Self { val, next: None }
    }
}

/*
Create the following singly linked list

4 -> 7 -> 3 -> 9 -> 1 -> 8
*/

pub fn sll_one_head() -> Option<Box<SingleLLNode>> {
    linked_list_from_slice(&[4, 7, 3, 9, 1, 8])
}

/*
Utils
*/

pub fn linked_list_from_slice(values: &[i32]) -> Option<Box<SingleLLNode>> {
    let mut head = None;

    for &value in values.iter().rev() {
        let mut node = Box::new(SingleLLNode::new(value));
        node.next = head;
        head = Some(node);
    }

    head
}

pub fn linked_list_to_vec(head: &Option<Box<SingleLLNode>>) -> Vec<i32> {
    let mut values = Vec::new();
    let mut current = head.as_ref();

    while let Some(node) = current {
        values.push(node.val);
        current = node.next.as_ref();
    }

    values
}

pub fn print_linked_list(head: &Option<Box<SingleLLNode>>) {
    let values: Vec<String> = linked_list_to_vec(head)
        .iter()
        .map(|value| value.to_string())
        .collect();

    println!("{}", values.join(" -> "));
}
